src/main.rs

```diff
@@ -314,13 +335,17 @@ fn main_send(
         }
 
         let state_clone = state_arc.clone();
+
         let push_thread = std::thread::spawn(move || {
+            let mut rl = RateLimiter::new(MAX_CHUNK_LEN, max_bandwidth_mbps);
```

Comment:
if we release the lock during sleep, another thread can show up and request the same amount of bytes.
I can implement the logic in a loop, but am not sure it's the best from a practical perspective: N threads will wake up in a loop and get blocked "forever" while only a subset of them make progress

what do you think?

Reply:
> N threads will wake up in a loop and get blocked "forever" while only a subset of them make progress

Yep, so now we get to the subtleties :grimacing: I have a feeling that we shouldn’t sleep while holding the lock because if a different thread needs to only read it to check something then it blocks. But at the same time I don’t think that is a problem right now and it would be the simplest thing to do.

If we don’t hold the lock while sleeping then we might get the thundering herd of threads all waking up but only one can consume the new available bytes. It’s only a few threads so I don’t expect overhead is a big problem, but it may lead to a fairness problem where one thread is winning all the time and the others are not sending. That is also not a problem in one sense, we’ll still send all the data, but it means we’re only utilizing one TCP connection. Which is maybe not a problem if we can saturate the target rate with one connection. But if we really want to do it properly, what we can do is add a `VecDeque` next to to the rate limiter, also protected by the lock, where we put the thread ids of waiting threads. Then if the limit is exhausted and you are not already in the queue, you put yourself at the end of the queue and go to sleep until you think the bucket is full enough already. When you wake up, if you are at the front of the queue then you consume from the limit, otherwise go back to sleep again.

It’s a lot more complex than sleeping with the lock, but sleeping with the lock does not guarantee fairness either because `std::sync::Mutex` does not guarantee fairness, so we can still have this situation where only one thread is sending all the time so we use only one TCP connection. Do we really need fairness though? Probably not but maybe we should measure this? Though that would be a lot of work and probably unfair is good enough …